### PR TITLE
Filter by App on User Aggregation Status Report

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

A new "Application" filter has been added to the User Aggregation Status report:
Screenshot from 2024-02-26 12-30-09

When no application filter is selected the report will show aggregation statistics for all applications (current behaviour).

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3412).

Currently the report simply looks at the last submission and sync dates for users when aggregating the statistics. This does not work when filtering by application however, as this data sits inside a nested ES field that contains the submission and sync dates for each application the mobile worker has used. As such, nested filtering and aggregation has been implemented to correctly retrieve the submission/sync date according to the filtered app ID. 

## Feature Flag

None.

### Safety story

- Local testing
- QA done

### Automated test coverage

A unit test is added to ensure that the `test_get_histogram_aggregation_for_app()` function returns the expected ES aggregates.

### QA Plan

QA completed. Ticket is available [here](https://dimagi.atlassian.net/browse/QA-6155).

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/reports/standard/deployments.py [hunk 1]
```diff
@@ -664,6 +668,26 @@ class AggregateUserStatusReport(ProjectReport, ProjectReportParametersMixin):
     def decorator_dispatcher(self, request, *args, **kwargs):
         super(AggregateUserStatusReport, self).decorator_dispatcher(request, *args, **kwargs)
 
+    @property
+    @memoized
+    def selected_app_id(self):
+        return self.request_params.get(SelectApplicationFilter.slug, None)
+
+    def _get_histogram_aggregation_for_app(self, field_name):
```
Comment (mkangia) [inline corehq/apps/reports/standard/deployments.py]: Hey @zandre-eng 
Can you help understand what is happening in this function?
For someone not familiar with nested aggregation, filter aggregation and data histogram, this function could use some documentation.

File: corehq/apps/reports/standard/deployments.py [hunk 2]
```diff
@@ -781,8 +781,18 @@ def _pct(val, total):
         query = self.user_query().run()
 
         aggregations = query.aggregations
-        last_submission_buckets = aggregations[0].normalized_buckets
-        last_sync_buckets = aggregations[1].normalized_buckets
+        if self.selected_app_id:
+            last_submission_buckets = (aggregations[0]
+                                       .filtered_last_submissions
+                                       .last_submissions_date_histogram
+                                       .normalized_buckets)
```
Comment (mkangia) [inline corehq/apps/reports/standard/deployments.py:788]: same as the other comment, can you help understand why this is different for if there is a selected app id?

File: corehq/apps/reports/standard/deployments.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/reports/standard/deployments.py]: As far as I know the `NestedAggregation` class is used to query nested data in ES (_"...this data sits inside a nested ES field"_), `FilterAggregation` is used to filter the specific key-value pair (_" correctly retrieve ... according to the filtered app ID"_) and DateHistogram (which I am not familiar with) I presume is to aggregate the "submission/sync date".

**Review by Charl1996 — COMMENTED** [review]
I'm happy with this PR subject to the answer of [this comment](https://github.com/dimagi/commcare-hq/pull/34178/files#r1502702934) of @mkangia

File: corehq/apps/reports/standard/deployments.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/reports/standard/deployments.py]: (the ES aggregation classes is hard to understand...can't say I have an unshakeable grasp on them)

Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: Charl is spot on with his explanation. The `DateHistogram` aggregation is used to group documents by a specific time interval. This allows us to aggregate all submission/sync dates that fall on the same dates and show these in a histogram graph on the UI.

It sounds like there is a bit of confusion around how this function works, so I'll look at adding a doc string to better explain what's going on.

File: corehq/apps/reports/standard/deployments.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py:788]: The aggregations we used are all nested inside each other in the following hierarchy (nested/filter/histogram). When creating these aggregations we assigned them each a name (first input param for the aggregation class). These names are where the results of the relevant aggregation will reside in `aggregations` of the ES query. Since the aggregations are nested inside each other, we have to drill down into each of them to eventually get to the buckets for the histogram.

TL:DR This is different because we make use of a nested and filter aggregation. Since all the aggregations are nested inside each other, we need to drill down all the aggregations to get to the buckets for the histogram.

File: corehq/apps/reports/standard/deployments.py (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: I have added a docstring to the function ([8bfe57a](https://github.com/dimagi/commcare-hq/pull/34178/commits/8bfe57a7d4007a3781c04577fb95411a8aa07d5f)). @mkangia @Charl1996 let me know if the docstring explains the function well enough, or if it needs any refactoring.

**Review by kaapstorm — APPROVED** [review]
Thanks for the helpful docstring. I always find it useful to see the JSON that's getting rendered.

**Review by Charl1996 — APPROVED** [review]
Looks good!

File: corehq/apps/reports/standard/deployments.py (hunk 1 above)
Comment (mkangia) [reply corehq/apps/reports/standard/deployments.py]: Thanks @zandre-eng 

It is a complicated function. I think the naming added to the challenge a bit.

File: corehq/apps/reports/standard/deployments.py [hunk 3]
```diff
@@ -663,6 +668,55 @@ class AggregateUserStatusReport(ProjectReport, ProjectReportParametersMixin):
     def decorator_dispatcher(self, request, *args, **kwargs):
         super(AggregateUserStatusReport, self).decorator_dispatcher(request, *args, **kwargs)
 
+    @property
+    @memoized
+    def selected_app_id(self):
+        return self.request_params.get(SelectApplicationFilter.slug, None)
+
+    def _get_histogram_aggregation_for_app(self, nested_field_name, date_field_name):
```
Comment (mkangia) [inline corehq/apps/reports/standard/deployments.py:676]: nit: I believe the parameter should be called "field_name" only. Its not clear to me why its called "nested_" already.

File: corehq/apps/reports/standard/deployments.py [hunk 4]
```diff
@@ -663,6 +668,55 @@ class AggregateUserStatusReport(ProjectReport, ProjectReportParametersMixin):
...
+                            "date_histogram": {
+                                ...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        """
+        nested_field_path = f'reporting_metadata.{nested_field_name}'
```
Comment (mkangia) [inline corehq/apps/reports/standard/deployments.py:706]: nit: same note here about naming. I believe it would be better to just call it "field_path"

File: corehq/apps/reports/standard/deployments.py (hunk 2 above)
Comment (mkangia) [reply corehq/apps/reports/standard/deployments.py:817]: okay, what are `filtered_last_submissions` & `last_submissions_date_histogram`?

File: corehq/apps/reports/standard/deployments.py (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py:676]: I called this `nested_field_name` to try and make it easy to identify that this is a nested field in the `users` ES index. Happy to change this if it is causing more confusion than it's helping.

Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: [9aaabbf](https://github.com/dimagi/commcare-hq/pull/34178/commits/9aaabbfb5ae31c4d87cc4f20fc293b69026b7043)

File: corehq/apps/reports/standard/deployments.py (hunk 4 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: [9aaabbf](https://github.com/dimagi/commcare-hq/pull/34178/commits/9aaabbfb5ae31c4d87cc4f20fc293b69026b7043)

File: corehq/apps/reports/standard/deployments.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py:817]: The `filtered_last_submissions` is the `FilterAggregation` and `last_submissions_date_histogram` is the `DateHistogram` that we added. Since the histogram aggregation is nested inside the filter aggregation which is nested inside the nested aggregation, we have to drilldown through the nesting to reach the histogram with `NestedAggregation.FilterAggregation.HistogramAggregation.normalized_buckets`

File: corehq/apps/reports/standard/deployments.py (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: I'm happy to change the function name if it's bringing additional confusion. @mkangia Any suggestions on what naming would be better?

Comment (mkangia) [reply corehq/apps/reports/standard/deployments.py]: Function naming is okay. It was the variable naming that got me which I highlighted in the other comment.

**Comment by mkangia** [comment]
Thanks for the naming changes @zandre-eng, those look good to me now.

File: corehq/apps/reports/standard/deployments.py (hunk 2 above)
Comment (mkangia) [reply corehq/apps/reports/standard/deployments.py:817]: > The filtered_last_submissions is the FilterAggregation and last_submissions_date_histogram is the DateHistogram that we added

Ah, it is coming from the dynamically generated naming we did.
This is confusing here because it looks like a property being referred when it is not, so searching for `filtered_last_submissions` and `last_submissions_date_histogram` in code does not give anything.

Can that be converted to a dict fetch like 
`aggregations[0]['filtered_last_submissions']`? Or could think of better ways like calling `_get_histogram_aggregation_for_app` with `filtered_last_submissions` and `last_submissions_date_histogram` or just leave a comment.

this is a non blocking comment.

**Review by mkangia — APPROVED** [review]

File: corehq/apps/reports/standard/deployments.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py:817]: Given that we use dot notation throughout HQ to refer to named aggregations (e.g. [here](https://github.com/dimagi/commcare-hq/blob/1eae8fc1cc3ae3c267b6501343f4a9ae348b6da9/corehq/apps/app_manager/dbaccessors.py#L534) and [here](https://github.com/dimagi/commcare-hq/blob/1eae8fc1cc3ae3c267b6501343f4a9ae348b6da9/corehq/apps/reports/analytics/esaccessors.py#L64)) I'm opted to rather not convert it to a dict fetch and keep it consistent with the rest of HQ. 

I'm happy to add a comment however, and explain where the references are coming from.

Comment (mkangia) [reply corehq/apps/reports/standard/deployments.py:817]: Thanks @zandre-eng , fine either way.

Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py:818]: @mkangia A comment has been added ([5a6c37a](https://github.com/dimagi/commcare-hq/pull/34178/commits/5a6c37a3abda8a8cee7217643ea15db89652f833)).

Comment (snopoke) [reply corehq/apps/reports/standard/deployments.py:818]: I haven't fully thought this through but my on the fly 2c is that the dynamic naming isn't necessary and you could do away with it entirely, and just use the names passed in to the `_get_histogram_aggregation_for_app`

**Review by snopoke — APPROVED** [review]

**Comment by zandre-eng** [comment]
> I haven't fully thought this through but my on the fly 2c is that the dynamic naming isn't necessary and you could do away with it entirely, and just use the names passed in to the _get_histogram_aggregation_for_app

@snopoke Are you referring to simply using `aggregations[0].last_submissions`? I gave that a try, but unfortunately it throws an `IndexError` which makes sense as the buckets sit further down the nesting. 

Looking around on HQ, I wasn't able to find an example of this being done. For curiosity's sake, I also tried giving all the aggregations the name passed into `_get_histogram_aggregation_for_app()` but it seems to be the same issue.

**Comment by snopoke** [comment]
> > I haven't fully thought this through but my on the fly 2c is that the dynamic naming isn't necessary and you could do away with it entirely, and just use the names passed in to the _get_histogram_aggregation_for_app
> 
> @snopoke Are you referring to simply using `aggregations[0].last_submissions`? I gave that a try, but unfortunately it throws an `IndexError` which makes sense as the buckets sit further down the nesting.
> 
> Looking around on HQ, I wasn't able to find an example of this being done. For curiosity's sake, I also tried giving all the aggregations the name passed into `_get_histogram_aggregation_for_app()` but it seems to be the same issue.

It definitely works but if you use the same agg name then it does look a bit weird:
```
- result.aggregations.nested_field.filtered_field.field_date_histogram)
+ result.aggregations.field.field.field)
```

But since only the top level agg needs to be unique you can use the field name for the nested agg and then static names for the others:

```python

@es_test(requires=[user_adapter], setup_class=True)
class TestQuery(SimpleTestCase):

    def _add_agg(self, query, field_name, date_field_name):
        field_path = f'reporting_metadata.{field_name}'
        nested_agg = NestedAggregation(field_name, field_path)
        filter_agg = FilterAggregation(
            "filtered_agg",
            filters.term(f'{field_path}.app_id', "123123")
        )
        histogram_agg = DateHistogram(
            "date_histogram",
            f'{field_path}.{date_field_name}',
            DateHistogram.Interval.DAY
        )
        agg = nested_agg.aggregation(filter_agg.aggregation(histogram_agg))
        return query.aggregation(agg)

    def test_aggs(self):
        query = UserES()
        query = self._add_agg(query, "last_submissions", "submission_date")
        query = self._add_agg(query, 'last_syncs', 'sync_date')
        print(query.pprint())
        res = query.run()
        print(res.aggregations)
        print(res.aggregations.last_submissions.filtered_agg.date_histogram)
        print(res.aggregations.last_syncs.filtered_agg.date_histogram)
```

Side note. A unit test for the agg part would be nice to have as well. It should be easy to extract it to a function that doesn't depend on the report.

**Review by SmittieC — COMMENTED** [review]
Looks fine to me, although I understand little of this area of the code. Best for someone with more insight to give the final approval

**Review by Charl1996 — APPROVED** [review]
Looks good to me.

File: corehq/apps/reports/standard/deployments.py [hunk 5]
```diff
@@ -749,8 +791,13 @@ def _pct(val, total):
         query = self.user_query().run()
 
         aggregations = query.aggregations
-        last_submission_buckets = aggregations[0].normalized_buckets
-        last_sync_buckets = aggregations[1].normalized_buckets
+        if self.selected_app_id:
+            (last_submission_buckets, last_sync_buckets) = [
+                agg.filtered_agg.date_histogram.normalized_buckets for agg in aggregations
+            ]
```
Comment (mkangia) [inline corehq/apps/reports/standard/deployments.py:797]: nit: I believe using for loop here makes it harder to read. May be better to keep it simpler like the else block.

**Review by mkangia — APPROVED** [review]
Thanks for the feedback @snopoke and the refactor @zandre-eng

**Comment by mkangia** [comment]
@zandre-eng good to inform QA to do a fresh pass after the refactor done.

File: corehq/apps/reports/standard/deployments.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/apps/reports/standard/deployments.py]: [65b3675](https://github.com/dimagi/commcare-hq/pull/34178/commits/65b36757a36579afce6f205999c07e1f2c8dad74)

**Review by mkangia — APPROVED** [review]
